### ISSUE TITLE
Memetic local-search budget is allocated by current score, never by expected training gain — Jin (2011) §5

### ISSUE THREAD

**Issue by stservice** [stSoftwareAU/NEAT-TS#3934]
Parent: #3919 — Jin (2011) surrogate-assisted evolution, fleet gap analysis.
Related: #3929 (archive), #3930 (surrogate). Distinct from #3915 (distillation) and #3918 (SWA).

**Priority: MEDIUM (GRQ).** Real, GRQ-relevant, but smaller and lower-risk than the evaluation-cost gaps.

## What is missing

Jin (2011) §5 covers surrogates inside *memetic* algorithms — where the
expensive thing is not only fitness evaluation but the local search attached to
each individual. The surrogate's job there is to decide **who gets refined**:
local search is a large fixed cost per individual, so spending it on individuals
that will not benefit is the dominant waste.

NEAT-AI is a memetic algorithm. Per-generation backpropagation is Lamarckian
local search, and `src/blackbox/MemeticTrajectory.ts` / `MemeticUpdate.ts`
already preserve pre-mutation state and record the weight/bias deltas that
training produced. The infrastructure for reasoning about local-search outcomes
is there.

The selection rule is not. `src/NEAT/TrainingCandidates.ts` is the whole policy:

```ts
// Pick up to `limit` training candidates from a score-sorted population.
```

It takes the top `trainPerGen` creatures by current score, skipping non-finite
scores. That is pure exploitation of *present* fitness, and it answers a
different question from the one that matters. "Who is currently best?" is not
"who will gain most from a gradient step?" — and on a converged lineage those
two sets can be nearly disjoint, because the current leader is likely to be the
individual closest to its local optimum and therefore the one with the least
left to extract.

Nothing in the codebase estimates expected training gain. The memetic
trajectories that would supply the training data for such an estimate are
recorded and used only for restoration.

## How this improves GRQ's 90-day predictions

Backpropagation on a 5,317-neuron / 38,988-synapse forward-only creature over a
21.2 GiB corpus is a substantial slice of a run that only completes one or two
generations in ~16 minutes. `trainPerGen` defaults to 20 % of the population for
supervised costs (`src/config/TrainPerGen.ts`), so a meaningful fraction of GRQ's
scarce wall-clock goes to gradient steps allocated by a rule that has never been
validated against outcomes.

The plausible win is a reallocation rather than a saving: the same training
budget spent on creatures with headroom instead of on the incumbent. Given
accepted improvements around 1e-05 on this lineage, moving even a couple of
gradient steps per generation onto candidates that still have slack is a
credible source of exactly that magnitude of gain.

It should be said plainly that this is the most speculative *mechanism* in the
sweep — but it is also the cheapest to test, because the answer is already
latent in the memetic trajectories the fleet has been recording all along.

## Proposed

Two stages, and the first is worth doing on its own.

**Stage 1 — measure the current rule. No new selection logic.**
Record, per training event: the creature's descriptor (#3929), its pre-training
score, its post-training score, and its rank in the population. Then answer:

- Does score-rank predict training gain at all? Rank-correlate pre-training rank
  against realised gain.
- Does the current top-`trainPerGen` rule beat random selection of the same
  size, on realised gain per unit of training wall-clock?

If the top-rank rule is no better than random, that is a finding worth having
regardless of whether any surrogate is ever built — and it is obtainable from
instrumentation alone.

**Stage 2 — only if Stage 1 shows gain is predictable.**
A gain predictor fitted to that record, used to select training candidates by
predicted *improvement* rather than by current score. Keep a floor: the current
fittest is always trained, so the rule can never starve the incumbent, and a
fraction of slots stay randomly allocated so the predictor keeps receiving data
from candidates it would otherwise never select.

## Scope boundary

This issue is about **who is selected for local search**. It is not about how
that local search runs internally — #3918 (stochastic weight averaging) and
#3915 (ensemble distillation) own that territory and stay out of scope here.

## Failure detection

- Stage 1 must report the random-selection baseline. Without it the comparison
  is unfalsifiable.
- Any Stage 2 selector must be A/B'd at the same seed against the current rule,
  judged on **final exact score**, not on mean training gain — a selector that
  maximises measured gain by picking creatures with the most room to improve can
  systematically favour poor creatures. That is the specific way this one goes
  wrong.
- Instrumentation must not measurably lengthen a training step; assert against a
  budget.
- A negative Stage 1 result closes the issue as a documented finding.

## Acceptance

- [ ] Per-training-event record: descriptor, pre/post score, population rank, wall-clock.
- [ ] Rank-vs-gain correlation reported over ≥200 real training events.
- [ ] Random-selection baseline reported alongside the current rule.
- [ ] Explicit go/no-go for Stage 2 recorded on #3919.
- [ ] If Stage 2 proceeds: current fittest always trained; a random slot fraction retained; both tested.
- [ ] Stage 2 A/B at the same seed judged on final exact score.
- [ ] No overlap with #3915 / #3918 scope.

## References

- Jin (2011), §5 — surrogate-assisted memetic algorithms and local-search budget allocation.
- Ong, Nair & Keane (2003), *Evolutionary optimization of computationally expensive problems via surrogate modeling* — surrogate-assisted local search in a memetic loop.
- `src/blackbox/MemeticTrajectory.ts`, `src/NEAT/TrainingCandidates.ts` (#2791), `src/config/TrainPerGen.ts` — the current mechanism and rule.

**Comment by stservice** [comment]
✅ **Vibe Coder released this claim** — host `vibe-coder-98063`, finished 05:55 UTC. Raised #4012 — https://github.com/stSoftwareAU/NEAT-AI/pull/4012
**Attempts on this issue:** 2
- 02:13 `vibe-coder-37099` — no PR (`timeout`, phase `execute`)
- 05:55 `vibe-coder-98063` — raised #4012

**Comment by stservice** [comment]
Claimed by `stservice-1789182158247` on host `vibe-coder-52841`

**Comment by stservice** [comment]
Automatic recovery: this issue was unassigned from `stservice` because the worker on machine `vibe-coder-2593-e43d0202-399a-4213-8ae6-0910910a3b22` appears to have stopped responding (no heartbeat for over 120 minutes) (last heartbeat 2026-09-12T03:13:03.000Z). The issue is now available to be picked up again on the next scan cycle.

---

Self-healing recovery (Issue #471)

**Comment by stservice** [comment]
✅ **Vibe Coder released this claim** — collapsed into the release summary above.

**Comment by stservice** [comment]
Pull request https://github.com/stSoftwareAU/NEAT-AI/pull/4012 has been created to address this issue.

**Comment by stservice** [comment]
## Issue run model stats

- **Requested model:** `opus`
- **Served model(s):** `claude-opus-5`
- **Effort:** `high`
- **Issue invocations:** 1
- **Tokens:** input 208 · output 68,624 · cache write 129,897 · cache read 20,499,527
- **Prompt cache:** 99.4% (read 20,499,527 · write 129,897 · uncached 208)
- **Turns:** 104
- **Duration:** 33m 27s
- **Estimated cost (USD, estimate only):** ~$12.78
  - `claude-opus-5`: $12.78 — input $0.0010 · output $1.72 · cache write $0.8119 · cache read $10.25
- **Degraded:** no

_Estimate only — this block covers the run that posted it. The issue total sums the run-stats comments visible on this issue; runs that reported no figures are not included._

**Comment by VibeCoderST** [comment]
Closed automatically — PR #4012 merged into `milestone/3919-surrogate-assisted-evolution-jin-2011-fleet-g` — a milestone branch, so the change reaches the default branch with that milestone's rollup — and its change landed (merge commit `492c40432fc6a072806aa13d3265afc603cfc478`, via `milestone-route-open`).

This issue was not closed by the run that produced the fix, so every claim scan refused it as `merged-pr-permanent` while it stayed open. The housekeeping merged-PR issue sweep closed it (Issue #504).

**Comment by VibeCoderST** [comment]
## Milestone progress: #3919 Surrogate-assisted evolution (Jin 2011): fleet gap analysis for GRQ gains

**10 of 10** issues complete — all issues in milestone '#3919 Surrogate-assisted evolution (Jin 2011): fleet gap analysis for GRQ gains' are now complete! A summary PR will be created shortly.
